### PR TITLE
Wrote select, search, treeSearch, treePosition, delete, insert, treeInsert, remove, removeLeaf, side

### PR DESCRIPTION
**Pull request by eytan-c**
Both are similar functions, but with different uses.
1) Select, finds the node in the tree, and returns an external node if the node doesn't exist.
2) findInsertParent - finds the node underwhich to insert a new node.
The only issue that might arise in insert is if we want to insert a node with a key that is already in the tree.
If we could return a tuple in Java that would solve it (and we can just have 1 function) but I don't know if that is possible.

### CODE REVIEW

**Comment by Onoam** [comment]
According to the project specifications, we don't need to insert a node if one already exists with its key. We just return -1 and do nothing.

File: WAVLTree/src/WAVLTree.java [hunk 1]
```diff
@@ -38,9 +38,63 @@ public boolean empty() {
...
+	 * If node doesn't exist, return external node.
+	 * Complexity - O(log n). Upper bounded by height of tree.
+	 * @param k int - key of node to select
+	 * @return current WAVLNode - the node with key k, or external node.
+	 */
+	private WAVLNode select(int k) {
+		WAVLNode current = getRoot();
+		WAVLNode parent;
+		while (current.getRank() != -1) {
+			if (k = current.getKey()) {
```
Comment (Onoam) [inline WAVLTree/src/WAVLTree.java]:
```suggestion
			if (k == current.getKey()) {
```

File: WAVLTree/src/WAVLTree.java [hunk 2]
```diff
@@ -38,9 +38,63 @@ public boolean empty() {
...
+	 * Finds the WAVLNode underwhich to insert a new WAVLNode with key k.
+	 * Complexity - O(log n). Upper bounded by height of tree.
+	 * @param k int - the new key that we want to insert
+	 * @return parent WAVLNode - the WAVLNode underwhich to insert.
+	 */
+	private WAVLNode findInsertParent(int k) {
+		WAVLNode current = getRoot();
+		WAVLNode parent;
+		while (current.getRank() != -1) {
+			if (k = current.getKey()) {
```
Comment (Onoam) [inline WAVLTree/src/WAVLTree.java]:
```suggestion
			if (k == current.getKey()) {
```

File: WAVLTree/src/WAVLTree.java [hunk 3]
```diff
@@ -38,9 +38,90 @@ public boolean empty() {
...
+	 * If this is legal, this can stop us from having 2 functions that do the same thing.
+	 * @param k
+	 * @return
+	 */
+	private WAVLNode[] findNode(int k) {
+		WAVLNode current = getRoot();
+		WAVLNode parent = current.getParent();
+		WAVLNode[] array = new WAVLNode[2];
+		while (current.getRank() != -1) {
+			if (k = current.getKey()) {
```
Comment (eytan-c) [inline WAVLTree/src/WAVLTree.java]:
```suggestion
			if (k == current.getKey()) {
```

**Comment by eytan-c** [comment]
@Onoam Let's decide on what functions to use for the finding.

**Comment by eytan-c** [comment]
Lot's of functions to go over. Some logic:
For the actual inserting and deleting, wrote private functions that do the work (for readability).
Also added some helper functions for different cases of delete.
Regarding the search etc, I decided to go with what was in the slides, but let's discuss what to keep out of all the options. #5 #7 #6

**Comment by eytan-c** [comment]
also, I assumed the signature for `rebalance` would be `private int rebalance(WAVLNode node)`

**Comment by Onoam** [comment]
This tons of code! I can't actually read all of it in detail, but from what I read it looks good, I'm just gonna assume that it all is. ;)

**Pull request merged**